torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -255,11 +271,47 @@ void Reducer::autograd_hook(VariableIndex index) {
       mark_variable_ready(unused_index);
     }
   }
-
   // Finally mark variable for which this function was originally called.
   mark_variable_ready(index);
 }
 
+void Reducer::delayed_autograd_hook() {
+  std::lock_guard<std::mutex> lock(this->mutex_);
+
+  // Ignore if we don't expect to be called.
+  // This may be the case if the user wants to accumulate gradients
+  // for number of iterations before reducing them.
+  if (!expect_autograd_hooks_) {
+    return;
+  }
+
+  if (require_final_hook_) {
+    torch::autograd::Engine::get_default_engine().queue_callback([=] {
```

Comment:
Is this guaranteed to run in the end? IIUC, the checkpoint API does end up calling re-entrant backward. In this case, won't the `final_callbacks_` hook get executed after a nested backward run is done instead of running after the entire backward pass is finished?

Reply:
I think this is fine. Those reentrant callbacks should finish before the original backward pass, as their return values are needed in the original backward pass. So, the final callback on the original backward pass should work.

Reply:
IIUC, `queue_callback` queues the callback functions on the engine and those callbacks are not associated with any backward pass. Once any backward call finishes, won't all the callbacks get executed here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L726?

Reply:
Ahh, yes, that's right. And I realized that the test didn't fail because I forgot to run the optimizer. Let me fix this.